src/suggestion-box.ts

```diff
@@ -0,0 +1,383 @@
+        }
+        return this.private_items.indexOf(el);
+    }
+    saveInputEvent(event: Event) {
+        if (this.result[this._indexOfTarget(event)]) {
+            (this.input_el as HTMLInputElement).value =
+                this.result[this._indexOfTarget(event)]?.["name"];
+        } else if (this._indexOfTarget(event) === this.result.length) {
+            this.loadmoreentry(event);
+        }
+    }
+
+    mouseOverEvent(e: any) {
+        this.curr_index = this._indexOfTarget(e);
+        this.index = this.curr_index;
+
+        if (this.prevIndex >= 0) {
+            const prevItem = this.private_items[this.prevIndex];
+            if (prevItem) {
+                prevItem.classList.remove("sgn-active");
+            }
+        }
+
+        if (this.curr_index >= 0) {
+            this.private_items[this.curr_index].classList.add("sgn-active");
+        }
+
+        this.prevIndex = this.curr_index;
+    }
+
+    inputfield() {
+        this.index = -1;
+    }
+    private _inputFocusEvent(event: any) {
+        if (event.target.tagName == "INPUT") this.show_container = true;
+    }
+    private _mouseOutEvent() {
+        if (this.prevIndex >= 0 && this.prevIndex < this.result.length) {
+            const prevItem = this.private_items[this.prevIndex];
+            if (prevItem) {
+                prevItem.classList.remove("sgn-active");
+            }
+        }
+    }
+    windowClickEvent(event: any) {
+        let el = event.target;
+        console.log(el.parentElement);
+        while (el) {
+            if (el == this) return;
+            el = el.parentElement;
+        }
+        this.show_container = false;
+    }
+    private _format(str: any, obj: any) {
+        if (typeof obj == "string" || typeof obj == "number") return obj;
+        else {
+            for (let key of Object.keys(obj)) {
+                let template: any;
+                if (key === "name") {
+                    template =
+                        this.highlight === "matched"
+                            ? this._highlight(obj[key])
+                            : this._startsWithHighlight(obj[key]);
+                } else {
+                    template = obj[key];
+                }
+
+                str = str.replace(
+                    new RegExp("\\{" + key + "\\}", "gi"),
+                    template
+                );
+            }
+            return str;
+        }
+    }
+    keyDownInputEvent(e: any) {
+        const resultLength = this.result.length;
+        switch (e.key) {
+            case "ArrowUp":
+                this.index =
+                    this.index === -1 || (this.has_more && this.index === 0)
+                        ? resultLength + (this.has_more ? 0 : 0)
+                        : (this.index -
+                              1 +
+                              resultLength +
+                              (this.has_more ? 1 : 0)) %
+                          (resultLength + (this.has_more ? 1 : 0));
```

Comment:
I beleive this condition be replaced to 
`this.index == -1?(resultLength + this.has_more?1:0):(this.index - 1)`